### PR TITLE
Allow power manager to disable charging

### PR DESCRIPTION
**Pull request by eberseth**
### Problem

The power manager may override user requests to disable battery charging.  While the user may set the charge state on the PMIC, the power manager may enable charge state silently.

### Solution

Add a flag for non-volatile power management configuration that ensures that the power manager honor user requests to disable charging.  Additionally create get methods for the configuration so that the user can query the current configuration.

### Steps to Test

Which unit/integration/application tests are applicable to this code change? (At minimum a test of some kind should be provided)

### Example App

```c
SerialLogHandler logHandler;

void setup() {
    // Apply a custom power configuration
    SystemPowerConfiguration conf;

    conf.feature(SystemPowerFeature::DISABLE_CHARGING);

    int res = System.setPowerConfiguration(conf);
    Log.info("setPowerConfiguration=%d", res);
    // returns SYSTEM_ERROR_NONE (0) in case of success

    // Settings are persisted, you normally wouldn't do this on every startup.
}

void loop() {
    auto config = System.getPowerConfiguration();
    Log.info("Current PMIC settings:");
    Log.info("PMIC detection: %s", cfg.isFeatureSet(SystemPowerFeature::PMIC_DETECTION) ? "true" : "false");
    Log.info("Use VIN with USB: %s", cfg.isFeatureSet(SystemPowerFeature::USE_VIN_SETTINGS_WITH_USB_HOST) ? "true" : "false");
    Log.info("Power manager disabled: %s", cfg.isFeatureSet(SystemPowerFeature::DISABLE) ? "true" : "false");
    Log.info("Charging disabled: %s", cfg.isFeatureSet(SystemPowerFeature::DISABLE_CHARGING) ? "true" : "false");
    Log.info("Input voltage: %u mV", cfg.powerSourceMinVoltage())
    Log.info("Input current: %u mA", cfg.powerSourceMaxCurrent());
    Log.info("Charge voltage: %u mV", cfg.batteryChargeVoltage());
    Log.info("Charge current: %u mA", cfg.batteryChargeCurrent());
}
```

### References

N/A

---

### Completeness

- [x] User is totes amazing for contributing!
- [ ] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [ ] Problem and Solution clearly stated
- [ ] Run unit/integration/application tests on device
- [ ] Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: system/inc/system_dynalib.h [hunk 1]
```diff
@@ -102,7 +102,8 @@ DYNALIB_FN(BASE_IDX + 16, system, system_invoke_event_handler, int(uint16_t hand
 
 #if HAL_PLATFORM_POWER_MANAGEMENT
 DYNALIB_FN(BASE_IDX + 17, system, system_power_management_set_config, int(const hal_power_config*, void*))
-#define BASE_IDX1 (BASE_IDX + 18)
+DYNALIB_FN(BASE_IDX + 18, system, system_power_management_get_config, int(hal_power_config*, void*))
+#define BASE_IDX1 (BASE_IDX + 19)
```
Comment (XuGuohui) [inline system/inc/system_dynalib.h:106]: This breaks the dynalib table, which will cause incompatibility. We should keep `BASE_IDX1` and append the dynalib function at the end of the table. So it should be:
```
#if HAL_PLATFORM_POWER_MANAGEMENT
DYNALIB_FN(BASE_IDX1 + 2, system, system_power_management_get_config, int(hal_power_config*, void*))
#define BASE_IDX2 (BASE_IDX1 + 3)
#else
#define BASE_IDX2 (BASE_IDX1 + 2)
#endif // HAL_PLATFORM_POWER_MANAGEMENT
```
following the `DYNALIB_FN(BASE_IDX1 + 1, system, system_reset, int(unsigned, unsigned, unsigned, unsigned, void*))`.

File: system/src/system_power_manager.cpp [hunk 2]
```diff
@@ -176,14 +190,16 @@ void PowerManager::handleUpdate() {
   // consecutively. The 1st reads fault register status from the last read and the 2nd
   // reads the current fault register status.
   volatile uint8_t lastFault = power.getFault();
-  (void)lastFault;
-  const uint8_t curFault = power.getFault();
+  const uint8_t curFault = power.getFault() | lastFault;
```
Comment (XuGuohui) [inline system/src/system_power_manager.cpp:193]: We didn't `|` the `lastFault`, any reason why we do it now?

File: system/src/system_power_manager.cpp [hunk 3]
```diff
@@ -176,14 +190,16 @@ void PowerManager::handleUpdate() {
...
 
   // Watchdog fault or buck converter got disabled
   if ((curFault & 0x80) || !power.isBuckEnabled()) {
```
Comment (XuGuohui) [inline system/src/system_power_manager.cpp:196]: Nit: magic number.

File: wiring/inc/spark_wiring_power.h [hunk 4]
```diff
@@ -102,7 +102,8 @@ class PMIC {
     byte readChargeTermRegister();
     bool disableWatchdog(void);
     bool setWatchdog(byte time);
-
+    bool enableSafetyTimer();
```
Comment (XuGuohui) [inline wiring/inc/spark_wiring_power.h:105]: Nit: return `int` to indicate errors.

**Review by XuGuohui — CHANGES_REQUESTED** [review]

**Comment by monkbroc** [comment]
I'll let Guo-Hui and Andrey review. This is beyond my knowledge to meaningfully review.

File: wiring/inc/spark_wiring_system.h [hunk 5]
```diff
@@ -721,6 +721,12 @@ class SystemClass {
         return system_power_management_set_config(conf.config(), nullptr);
     }
 
+    particle::SystemPowerConfiguration getPowerConfiguration() {
+        hal_power_config config {0};
```
[the remark this answers is not in the document]
Comment (eberseth) [reply wiring/inc/spark_wiring_system.h:725]: I'll use that then.

File: user/tests/wiring/power_management/power.cpp [hunk 6]
```diff
@@ -530,6 +594,20 @@ test(POWER_15_DisableFeatureFlag) {
...
+        assertEqual(cfg.isFeatureSet(SystemPowerFeature::DISABLE), true);
+        assertEqual(cfg.isFeatureSet(SystemPowerFeature::DISABLE_CHARGING), false);
+        assertEqual(cfg.getPowerSourceMaxCurrent(), particle::power::DEFAULT_INPUT_CURRENT_LIMIT);
+        assertEqual(cfg.getPowerSourceMinVoltage(), particle::power::DEFAULT_INPUT_VOLTAGE_LIMIT);
+        assertEqual(cfg.getBatteryChargeCurrent(), particle::power::DEFAULT_CHARGE_CURRENT);
+        assertEqual(cfg.getBatteryChargeVoltage(), particle::power::DEFAULT_TERMINATION_VOLTAGE);
+
         assertEqual(System.batteryState(), (int)BATTERY_STATE_UNKNOWN);
         assertEqual(System.powerSource(), (int)POWER_SOURCE_UNKNOWN);
         // Restore default power management configuration
```
Comment (avtolstoy) [inline user/tests/wiring/power_management/power.cpp:613]: TODO: (can be a separate PR) let's add three tests 1) disable/enable charging in runtime and verifying configuration and actual PMIC state 2) disable/enable charging, resetting and making sure that the necessary charging state is set on boot 3) disable/enable chargin, triggering PMIC watchdog and making sure that the necessary charging state is achieved after watchdog timer expires.

**Review by avtolstoy — CHANGES_REQUESTED** [review]
Mainly due to dynalib change and ABI problem with `getConfig()`.

File: wiring/src/spark_wiring_power.cpp [hunk 7]
```diff
@@ -423,9 +423,11 @@ void PMIC::reset() {
 bool PMIC::enableCharging() {
     std::lock_guard<PMIC> l(*this);
     byte DATA = readRegister(POWERON_CONFIG_REGISTER);
-    DATA = DATA & 0b11001111;
-    DATA = DATA | 0b00010000;
-    writeRegister(POWERON_CONFIG_REGISTER, DATA);
+    if ((DATA & 0b00110000) != 0b00010000) {
```
[the remark this answers is not in the document]
Comment (eberseth) [reply wiring/src/spark_wiring_power.cpp:426]: So much magic that it has it's own Vegas show.

File: system/inc/system_dynalib.h (hunk 1 above)
Comment (eberseth) [reply system/inc/system_dynalib.h:106]: Good catch!  I thought it was already at the end

File: user/tests/wiring/power_management/power.cpp (hunk 6 above)
Comment (eberseth) [reply user/tests/wiring/power_management/power.cpp:613]: Yeah, I'm still testing some of this code so I didn't add the tests here yet.

File: system/src/system_power_manager.cpp [hunk 8]
```diff
@@ -675,44 +699,55 @@ int PowerManager::setConfig(const hal_power_config* conf) {
   return ret;
 }
 
-void PowerManager::loadConfig() {
+int PowerManager::getConfig(hal_power_config* conf) {
   // Ignore errors since we are anyway sanitizing the config and setting some sane defaults
-  config_.size = sizeof(config_);
-  int err = hal_power_load_config(&config_, nullptr);
+  conf->size = sizeof(*conf);
```
[the remark this answers is not in the document]
Comment (eberseth) [reply system/src/system_power_manager.cpp:704]: I can add an extra argument to specify target size `int PowerManager::getConfig(hal_power_config* conf, size_t size)`

File: system/src/system_power_manager.cpp [hunk 9]
```diff
@@ -675,44 +699,55 @@ int PowerManager::setConfig(const hal_power_config* conf) {
...
 
   // Sanitize configuration
-  if (err || config_.version == 0xff || !isValid(config_.size)) {
+  if (err || conf->version == 0xff || !isValid(conf->size)) {
     // Flags require special processing even if the config is not valid
     // in order to keep DCT compatibility for the HAL_POWER_PMIC_DETECTION feature flag
-    uint32_t flags = config_.flags;
-    config_ = defaultPowerConfig;
+    uint32_t flags = conf->flags;
+    *conf = defaultPowerConfig;
```
[the remark this answers is not in the document]
Comment (eberseth) [reply system/src/system_power_manager.cpp:712]: `memcpy(conf, &defaultPowerConfig, size)`?  `std::min()`?

[the remark this answers is not in the document]
Comment (eberseth) [reply system/src/system_power_manager.cpp:712]: if check `.size` of each?

File: system/src/system_power_manager.cpp (hunk 3 above)
Comment (eberseth) [reply system/src/system_power_manager.cpp:196]: Already there.  I'll add a comment for this case.

File: system/src/system_power_manager.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (avtolstoy) [reply system/src/system_power_manager.cpp:704]: There is no point in an extra argument. The structure itself already has a size field. It's a convention that the caller needs to specify the size it has allocated (which in our situation may be less than the current size of this structure, again given the fact that the app may be built for 1.0.0 and the system part currently on the device is already 2.0.0 where the structure has been extended).

The callee does not modify the size and uses it as a reference to know how much it can read or write out of the provided structure.

File: system/src/system_power_manager.cpp (hunk 2 above)
Comment (eberseth) [reply system/src/system_power_manager.cpp:193]: Correct

File: wiring/inc/spark_wiring_system_power.h [hunk 10]
```diff
@@ -42,33 +43,54 @@ class SystemPowerConfiguration {
...
     SystemPowerConfiguration(SystemPowerConfiguration&&) = default;
+    SystemPowerConfiguration(const hal_power_config& conf) : conf_(conf) {}
     SystemPowerConfiguration& operator=(SystemPowerConfiguration&&) = default;
 
     SystemPowerConfiguration& powerSourceMinVoltage(uint16_t voltage) {
         conf_.vin_min_voltage = voltage;
         return *this;
     }
 
+    uint16_t getPowerSourceMinVoltage() {
```
[the remark this answers is not in the document]
Comment (eberseth) [reply wiring/inc/spark_wiring_system_power.h:54]: Good catch

File: system/src/system_power_manager.cpp [hunk 11]
```diff
@@ -675,44 +699,55 @@ int PowerManager::setConfig(const hal_power_config* conf) {
...
-        config_.flags |= HAL_POWER_PMIC_DETECTION;
+        conf->flags |= HAL_POWER_PMIC_DETECTION;
       }
     }
 #else
     (void)flags;
 #endif // HAL_PLATFORM_POWER_MANAGEMENT_OPTIONAL
+
+    // This error code may be informative only since the configuration is corrected
+    err = SYSTEM_ERROR_INVALID_STATE;
```
[the remark this answers is not in the document]
Comment (eberseth) [reply system/src/system_power_manager.cpp:724]: Just was thinking that it would be nice to indicate something has been automatically corrected.  I'll take it out.

File: system/src/system_power_manager.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (avtolstoy) [reply system/src/system_power_manager.cpp:712]: Yes, something along the lines of `std::min` and `memcpy` limiting by that. See also https://github.com/particle-iot/device-os/pull/2257#discussion_r546898943 on the suggestion to simplify this a bit and only worry about limiting the size when getting the configuration out to an external caller that may require this.

File: system/src/system_power_manager.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (eberseth) [reply system/src/system_power_manager.cpp:704]: How did `config_ = defaultPowerConfig;` work in the original code before my changes?

[the remark this answers is not in the document]
Comment (avtolstoy) [reply system/src/system_power_manager.cpp:704]: @eberseth There is a difference if we are within 1 module, or something flows across the dynalib boundary. We don't need to worry about the sizes or anything if we are within the same module. The size of this structure is universal, because its definition was the same when building such module.

When calling something through dynalib it's not guaranteed. The application allocates the structure based on the definition it has in its version of Device OS, which may not have some of the changes present in newer version, but the system part may already have them.

```c++
// 1.0.0
struct SomeStruct {
    uint16_t size;
    uint32_t field1;
    uint32_t field2;
    uint32_t field3;
    uint32_t field4;
}; // 18 bytes

// 2.0.0
struct SomeStruct {
    uint16_t size;
    uint32_t field1;
    uint32_t field2;
    uint32_t field3;
    uint32_t field4;

    uint32_t field5;
    uint32_t field6;
    uint32_t field7;
    uint32_t field8;
}; // 34 bytes

// 1.0.0 application will allocate only 18 bytes based on its definition available in its sources and will indicate that the struct passed through dynalib is indeed only 18 bytes
// 2.0.0 system code will need to be careful with the size when both reading and writing to such structs that come through dynalib, as its own definition of this same struct has already grown to 34 bytes
```

File: system/src/system_power_manager.cpp [hunk 12]
```diff
@@ -162,6 +165,17 @@ void PowerManager::wakeup() {
   }
 }
 
+void PowerManager::handleDisable() {
```
[the remark this answers is not in the document]
Comment (eberseth) [reply system/src/system_power_manager.cpp]: Done

File: system/src/system_power_manager.cpp [hunk 13]
```diff
@@ -399,11 +417,17 @@ void PowerManager::initDefault(bool dpdm) {
...
   // Set recharge threshold to default value - 100mV
-  power.setRechargeThreshold(100);
+  power.setRechargeThreshold(PMIC_NORMAL_RECHARGE_THRESHOLD);
   power.setChargeCurrent(config_.charge_current);
 
   // Enable charging
-  power.enableCharging();
+  if ((config_.flags & HAL_POWER_CHARGE_STATE_DISABLE) == 0) {
+    power.enableCharging();
+    power.enableSafetyTimer();
```
[the remark this answers is not in the document]
Comment (eberseth) [reply system/src/system_power_manager.cpp]: Done

File: wiring/inc/spark_wiring_system.h [hunk 14]
```diff
@@ -721,6 +721,13 @@ class SystemClass {
         return system_power_management_set_config(conf.config(), nullptr);
     }
 
+    particle::SystemPowerConfiguration getPowerConfiguration() {
```
Comment (avtolstoy) [inline wiring/inc/spark_wiring_system.h:724]: `const`

**Review by avtolstoy — APPROVED** [review]

File: system/src/system_power_manager.cpp [hunk 15]
```diff
@@ -675,44 +695,59 @@ int PowerManager::setConfig(const hal_power_config* conf) {
   return ret;
 }
 
-void PowerManager::loadConfig() {
+int PowerManager::getConfig(hal_power_config* conf) {
+  auto dest_size = conf->size;
```
[the remark this answers is not in the document]
Comment (eberseth) [reply system/src/system_power_manager.cpp:699]: The code is not consistent on any particular coding style.  I'll make this change

**Review by XuGuohui — APPROVED** [review]
